### DISCUSSION TITLE
Get current tiddler text in JS macro

### DISCUSSION THREAD

**Discussion by danfarrow** [TiddlyWiki/TiddlyWiki5#6720] [Q&A] [answered]
I’ve made a macro which converts a wikitext table into a multiline sparkline-like graph. Currently the macro takes the wikitext as a parameter and parses it, but I would like it to parse the current tiddler text directly.

This feels like it is an extremely dumb question but I have searched here and in the Google group yet still cannot figure out how to access the text of the current tiddler from a JS macro.

~~Is it even possible?!~~

**Comment by pmario** [comment]
> ... multiline sparkline-like graph

Can you tell what a "multiline sparkline-like graph" is? ... A link would be nice. 

>Currently the macro takes the wikitext as a parameter and parses it, but I would like it to parse the current tiddler text directly.

I think it's not the best approach to use a macro for this uscase. Macros per definition are "text substitutions". So they take some input and return some text output. ... They don't have a mechanism to participate in the global "refresh" mechanism. 

My gut feeling tells my, you may be better of with a widget, that takes a "tiddler title" as an input and renders some HTML output.

**Comment by danfarrow** [comment]
Thanks @pmario. This is my graph for tracking typing speed & accuracy, as an example 😄
![image](https://user-images.githubusercontent.com/236580/173229012-daf97766-feb0-4b9d-b54f-06093c45cfef.png)
The way it currently works is as a JS macro which accepts a parameter of wikitext representing a table of data e.g.
```
|!Date|!WPM|!Accuracy|
|12/06|88|98.43%|
|11/06|90|96.98%|
|10/06|88|96.93%|
|09/06|85|98.15%|
|08/06|89|100%|
|07/06|84|99.76%|
|06/06|81|94.65%|
...
```
The macro creates an SVG image of the results and returns it, along with the wikitext input, so the result in the tiddler is an SVG graph with the table of data beneath.

I update the tiddlers regularly with new results added to the top of the wikitext table but what I’d prefer is for the macro to parse the text of the tiddler it’s used within and add a sparkline to any wikitext tables.

You may be right that a widget is a better solution! I just discovered [this tutorial](https://btheado.github.io/tw-widget-tutorial/) which may help.

I might also just stick with my existing solution which, although not the most elegant, works for my purposes!

**Comment by pmario** [comment]
It is possible to read the tiddler content by name. eg: If you write this to the F12 console in the dev tools `var ht = $tw.wiki.getTiddler("HelloThere") ` you'll get back an object.

`ht.fields.text` will contain the text of the tiddler. BUT ... It also will contain your macro code, which will make parsing more complicated. Especially if there is more text in the tiddler. ... That was my main concern. 

```
<<my-macro>>

|!Date|!WPM|!Accuracy|
|12/06|88|98.43%|
|11/06|90|96.98%|
```

See the screenshot for a way to debug and get all the info you probably want ;) I'm using FireFox for debugging, since it displays a nice structure similar to the repo. Chrome is possible too, but it doesn't show the nice directory like structure

If you need a second run. Just click the Run Button and add a "space" char to the text. 
It will trigger the breakpoint again

Hope that helps

Have fun!

![image](https://user-images.githubusercontent.com/374655/173230551-c3edf41b-5d1f-4aa9-8a58-cbfe23c0f8ad.png)

**Reply by danfarrow** [reply]
Thanks @pmario, that’s a great approach to debugging which should speed things up for me! I’m going to accept @Jermolene’s answer as it seems less hacky but tbh I am most likely to continue with my hacky approach!

**Comment by Jermolene** [comment]
Hi @danfarrow as @pmario notes, macros are simple functions with string parameters and no side effects, and have no intrinsic ability to interact with the refresh mechanism. They are intended to be the simplest possible interface for simple textual manipulation; in practice developers bend them to do more than that, but there are always problems as a result.

The easiest way to approach your application would be to make a custom parser for a mime type like `text/x-sparkline`. The parser accepts a block of text and returns a parse tree (aka an AST or abstract syntax tree). You may also need to implement a custom widget; your parser would output a parse tree containing an invocation of the widget, passing the data as attributes.

The railroad plugin is a good example of this approach: https://github.com/Jermolene/TiddlyWiki5/tree/master/plugins/tiddlywiki/railroad

**Reply by danfarrow** [reply]
Thanks @Jermolene - I need to get a better understanding of widgets before embarking on this approach but that is something I will be doing in future. Thanks for your suggestions!

**Comment by danfarrow** [comment, marked as answer]
I have now got this working, building on the previous answers - thanks again!

Inside the macro, this gets the text of the current tiddler:
```
const currentTiddlerName = this.variables.currentTiddler.value;
const currentTiddler = this.wiki.getTiddler(currentTiddlerName);
const text = currentTiddler.fields["text"];
```
I extract the table wikitext using a regular expression (only supports a single table currently):
```
const regex = /\|[^]*\|/g;
return regex.exec(text)[0];
```
